padertorch/ops/stft_module.py

```diff
@@ -0,0 +1,270 @@
+
+        window = _get_window(
+            window=window,
+            symmetric_window=symmetric_window,
+            window_length=self.window_length,
+        )
+        assert fading in [None, True, False, 'full', 'half'], fading
+        self.fading = fading
+        self.pad = pad
+        self.stft_kernel = get_stft_kernel(size, window)
+        self.istft_kernel_real, self.istft_kernel_imag = get_istft_kernel(
+            size, shift, window)
+
+    def __call__(self, inputs):
+        """
+        Args:
+            inputs: shape: [..., T], T is #samples
+            num_samples, list or tensor of #samples
+
+        >>> mixture = torch.rand((2, 6, 203))
+        >>> torch_stft = STFT(512, 20, window_length=40,\
+                              complex_representation='concat')
+        >>> torch_stft_out = torch_stft(mixture)
+        >>> torch_stft_out.shape
+        torch.Size([2, 6, 12, 514])
+        >>> from paderbox.transform import stft
+        >>> stft_out = stft(mixture.numpy(), 512, 20, window_length=40)
+        >>> stft_signal = np.concatenate(\
+                [np.real(stft_out), np.imag(stft_out)], axis=-1)
+        >>> np.testing.assert_allclose(torch_stft_out, stft_signal, atol=1e-5)
+
+        """
+
+        org_shape = inputs.shape
+        stride = self.shift
+        length = self.window_length
+        x = inputs.view((-1, org_shape[-1]))
+        # Pad with zeros to have enough samples for the window function to fade.
+        assert self.fading in [None, True, False, 'full', 'half'], self.fading
+        if self.fading not in [False, None]:
+            if self.fading == 'half':
+                pad_width= (
+                    (self.window_length - self.shift) // 2,
+                    ceil((self.window_length - self.shift) / 2)
+                )
+            else:
+                pad_width = self.window_length - self.shift
+                pad_width = (pad_width, pad_width)
+            x = F.pad(x, pad_width, mode='constant')
+
+        if self.pad:
+            if x.shape[-1] < length:
+                pad_size = length - x.shape[-1]
+                x = F.pad(x, (0, pad_size))
+            elif stride != 1 and (x.shape[-1] + stride - length) % stride != 0:
+                pad_size = stride - ((x.shape[-1] + stride - length) % stride)
+                x = F.pad(x, (0, pad_size))
+
+        x = torch.unsqueeze(x, 1) # [..., 1, T]
+        weights = self.stft_kernel.to(x)
+        encoded = F.conv1d(x, weight=weights, stride=stride)
+
+        encoded = encoded.view(*org_shape[:-1], *encoded.shape[-2:])
+        encoded = rearrange(encoded, '... feat frames -> ... frames feat')
+        encoded = torch.chunk(encoded, 2, dim=-1)
+        if self.complex_representation == 'stacked':
+            encoded = torch.stack(encoded, dim=-1)
+        elif self.complex_representation == 'concat':
+            encoded = torch.cat(encoded, dim=-1)
+        else:
+            raise NotImplementedError
```

Comment:
Is this good code style? I honestly don't know if it is correct to raise the `NotImplementedError` class. I frequently see this, but my intuition says that it should be `NotImplementedError()`. I also prefer it if the error message has some explanatory text. Even if the Exception should never be raised in this code, you can never know.

Btw, for `complex_representation != 'complex'`, a `ValueError` might be more appropriate.

Reply:
This error should never occour because a new complex_representation should already be catched in the init. This can only happen if somebody overwrites ```STFT.complex_representation``` after initialization. However, I will add a sentence as explanation to the error message.

Reply:
Yes, I know that this "should never happen". But you never know what the user will be doing...

Reply:
I think it should be at least `NotImplementedError(self.complex_representation)`. Just in case, that this exception is thrown  the user gets the necessary information about the internal state.

Reply:
done